### ISSUE TITLE
Create CONTRIBUTING.md for web app scripts

### ISSUE THREAD

**Issue by fenryxo** [tiliado/nuvolaplayer#128]
It is necessary to make CONTRIBUTING.md to outline best practices. Draft follows. See also tiliado/nuvola-app-jango#3, tiliado/nuvola-app-this-is-my-jam#2 and tiliado/nuvola-app-amazon-cloud-player#3.

---

**Draft of file CONTRIBUTING.md** that will be included in each repository.

---

Contributing to Happy Songs script
============================

:+1::tada: It seems you would like to contribute to the Happy Songs script for Nuvola Player.
Great to have you here. Here are a few ways you can help make this script better! :tada::+1:

Bug Reporting
-------------

Whenever you encounter a bug in Nuvola Player, don't hesitate to report it. First of all, it is necessary to submit
your issue to the right place:

* Contact [Tiliado support](mailto:support@tiliado.eu) to report issues with Tiliado.eu website (tiliado.eu domain), your Tiliado account and memberships.
  
* Report issues with Tiliado Repository installer and installation of Nuvola Player at
  [the Nuvola Player Users mailing list](https://groups.google.com/d/forum/nuvola-player-users).
  
* Use [the Nuvola Player Users mailing list](https://groups.google.com/d/forum/nuvola-player-users) also to ask general  questions about Nuvola Player.

* Finally, for reporting issues and bugs of the Nuvola Player application and supported services, follow  [the Bug Reporting Guidelines](https://github.com/tiliado/nuvolaplayer/wiki/Bug-Reporting-Guidelines).

Development
-----------------
 
If you would like to contribute some code to this repository, welcome on a board and read on! :+1:

### Documentation

* See [Service Integration Development website](http://tiliado.github.io/nuvolaplayer/development/apps.html)
* or ask a question at [the nuvola-player-devel mailing list](https://groups.google.com/d/forum/nuvola-player-devel).

### Developers without a commit access to this repository

* Clone/fork the master branch via GitHub or `git clone`.
* Don't work on the master branch directly, but create a feature/topic/devel branch via `git checkout -b branch-name`. Use meaningful branch name.
* Make your commits and follow guidelines for commit messages [LINK TO GUIDELINES].
* Push your feature branch to GitHub.
* Create a pull request following guidelines for pull requests [LINK TO GUIDELINES].

### Developers with a commit access to this repository
 
  * Keep the `master` branch clean and functional, because it's the stable/production/release branch.
  * For one-commit changes without a need of code review, work directly on the master branch. Follow guidelines for commit messages [LINK TO GUIDELINES].
  * For large changes, use feature/topic/devel branches. You can experiment and break things in these branches, but they must be stable and functional when they are about to be merged back to master. Follow guidelines for commit messages [LINK TO GUIDELINES].
  * Create a GitHub pull request following guidelines for pull requests [LINK TO GUIDELINES] if you would like to have a code review (preferred). Otherwise, feel free to perform a merge on your own following guidelines for merge commits [LINK TO GUIDELINES].

------

**Draft of guidelines** that will be incorporated into [Service Integration Guidelines](http://tiliado.github.io/nuvolaplayer/development/apps/guidelines.html) in Nuvola Player developer's documentation.

-----

### Commit Messages

Commit messages should follow this template:

```
Short (50 chars or less) summary of changes

More detailed explanatory text, if necessary.  Wrap it to
about 72 characters or so. The blank line separating the
summary from the body is critical (unless you omit the body
entirely); tools like rebase can get confused if you run
the two together.

Further paragraphs come after blank lines. Also don't
forget to link to issues.
    
Issue: tiliado/nuvolaplayer#128
```

* Write useful commit summary (the first line. max 50 chars):
    + Use verbs and don't end summary with a dot, e.g. "Add test for AboutDialog" instead of "Added test for AboutDialog." or "Adding test for AboutDialog."
    + Be specific, e.g. "Increase version to 2.1" instead of "Update metadata", "Remove extra comma to fix JSON parse error" instead "Fix JSON file".
    + You can use shortcuts "w/" = with, "w/o" = without, etc.
* If short summary is not clear enough, add one or a few paragraphs with description about what you changed and why.
* Add link to a Github issue, e.g. tiliado/nuvolaplayer#128, if there is any. See [Writing on GitHub](https://help.github.com/articles/writing-on-github/#references) for details.

### Pull Requests

If you would like to have your code contributions merged to the main repository, [create a pull request](https://help.github.com/articles/creating-a-pull-request/). Pull request should follow similar template like commit messages, because a merge commit message will be based it:

```
Short (50 chars or less) summary of changes

More detailed explanatory text, if necessary.  Wrap it to
about 72 characters or so. The blank line separating the
summary from the body is critical (unless you omit the body
entirely); tools like rebase can get confused if you run
the two together.

Further paragraphs come after blank lines. Also don't
forget to link to issues.
    
- Author: Pull Request Author Name <your@email>
- Reviewed by: FIXME <FIXME>
- Issue: tiliado/nuvolaplayer#128

---

If you have some other notes regarding the pull request,
add a separator "---" and then write anything you want to ;-)
This text won't be included in the merge commit message.
```

* The same rules as for the commit summary apply to **the title** of a pull requests.
* Always provide **description** of the pull request: what you changed and why. The merge commit message will be based on this description.
* Add lines:
   +  `Author: Pull Request Author Name <your@email>` - you will be recorded as the author of the merge commit (see bellow)
   + `Reviewed by: FIXME <FIXME>` - will be filled by a reviewer
 * Add link to a related issue, if there is any: `Issue: tiliado/nuvolaplayer#128`

### Merge Commits

If you are a maintainer of a repository, follow these rules how to accept pull requests and create merge commits. First of all, check whether the title of a pull request and a description conforms guidelines above. If not, notify the author of the pull request (less work for you) or prepare a well formed commit message on your own (more work for you).

#### Commit message

Message of merge commits should follow this template:

```
Short (50 chars or less) summary of changes

More detailed explanatory text, if necessary.  Wrap it to
about 72 characters or so. The blank line separating the
summary from the body is critical (unless you omit the body
entirely); tools like rebase can get confused if you run
the two together.

Further paragraphs come after blank lines. Also don't
forget to link to issues.

 - Reviewed by: Your Name <your@email>
 - Reviewed by: Another Reviewer <his@email>
 - Pull Request: fenryxo/test#2
 - Issue: fenryxo/test#3
```

* Link to a pull request `https://github.com/fenryxo/test/pull/2` → ` - Pull request: fenryxo/test#2`
* Link to an issue `https://github.com/fenryxo/test/issues/3` -> `+ - Issue: fenryxo/test#3`

#### No merge via GitHub

Don't use Github to do a merge. It's a pure crap:
 -  It creates a commit summary like "Merge pull request #1 from fenryxo/mybranch.". That has zero information value.
 - It provides only a basic text box to fill in a commit description. No control of line wrapping at 75 characters. 

![no_github_merge](https://cloud.githubusercontent.com/assets/853706/8862021/fb68290e-318d-11e5-8e5e-5989d7df0a83.png)

#### Merge via command line

 * Add remote repository if necessary: `git remote add gh-USER https://github.com/USER/REPO.git; git fetch gh-USER`
 * Switch to the branch of the pull request: `git checkout gh-USER/BRANCH`
 * Review and test changes.
 * Switch to the master branch: `git checkout master` 
 * Use `git merge --no-ff --no-commit gh-USER/BRANCH` instead of a plain `git merge` to merge the pull request.
 * Use `git commit --author "Pull Request Author Name <author@email>"` to commit the merge with a very descriptive commit message:
    * Fill in  `Reviewed by: Your Name <FIXME>`
    * Link to a pull request `https://github.com/fenryxo/test/pull/2` → ` - Pull Request: fenryxo/test#2`
    * Link to an issue `https://github.com/fenryxo/test/issues/3` -> `+ - Issue: fenryxo/test#3`
 * Optionally, you can squash all commits of the pull request and the merge commit into one final commit. However, you should not ask author of the pull request to do rebasing and squashing to keep entrance barrier for contributors as low as possible.
 * Finally, ``git push``.
 * Optionally, you can remove remote repository: `git remote remove gh-USER/BRANCH`

#### Releases

  * Create a Github issue "Make release X.Y" and assing it to @fenryxo when ready for a new release, but don't increase version fields in metadata.
  * Don't hesitate to release often. Packages of service integration script are lightweight
    with a typical build time a few seconds thanks to [fpm](https://github.com/jordansissel/fpm).
  * Users will love you if you release a fix as soon as possible.

**Comment by fenryxo** [comment]
Dear Tiliado developers, @SteVwonder, @skyghis, @Tar-Dingens, @thenyeguy, @ajabot, @s83, @eurythmia, @thepisu and @jnv. Let's discuss a draft of new contributor guidelines! What do you think?

**Comment by SteVwonder** [comment]
Looks good. The merge instructions were easy to follow.

Do you have anything against developers working in their own forks then sending pull requests back to the main repo (like developers without commit access would do)?

**Comment by fenryxo** [comment]
> Do you have anything against developers working in their own forks then sending pull requests back to the main repo (like developers without commit access would do)?

What's the benefit of this?

**Comment by skyghis** [comment]
I have an "issue" with the release workflow. After a commit/merge into master the version number will not match the code source version.

Suggestion: 
New branch is created and version number is changed in the code before any other edit.
Edits are made in this branch and, when ready, the branch is merged into master and a tag is added.
This way, master is always the last stable version and version number in sources matches released version number

**Comment by fenryxo** [comment]
> After a commit/merge into master the version number will not match the code source version.

That isn't an issue. Once the code from a feature branch is merged to the master branch, it is considered "stable and ready for release", but not "released" yet. More commits can follow before a new tagged release is made with incremented version number. Version number in side branches is irrelevant.

> New branch is created and version number is changed in the code before any other edit.

No, please. Feature commits cannot touch version number, only release commits can.

**Comment by fenryxo** [comment]
>> Do you have anything against developers working in their own forks then sending pull requests back to the main repo (like developers without commit access would do)?

> What's the benefit of this?

@SteVwonder: I mean, you can create pull request also from you local git clone of the main repo without a need to create a fork at GitHub. Just create a feature branch, push it to the main repo and make a pull request to the master branch.

**Comment by eurythmia** [comment]
I like the workflow. I am, however, confused about the release numbers. Are these version numbers generated when you build packages with fpm, or is there some other mechanism for incrementing the release number?

**Comment by fenryxo** [comment]
The release numbers are incremented manually. The build script uses them then to detect whether to build new packages.

**Comment by SteVwonder** [comment]
@fenryxo, sounds good.

**Comment by fenryxo** [comment]
I've made some changes to this section. Note that I'm still learning Git (I'm a former Bazaar user), so any comments and suggestions are welcome ;-)

#### Merge via command line

 * Add remote repository if necessary: `git remote add gh-USER https://github.com/USER/REPO.git; git fetch gh-USER`
 * Switch to the branch of the pull request: `git checkout gh-USER/BRANCH`
 * Review and test changes.
 * Switch to the master branch: `git checkout master` 
 * Use `git merge --no-ff --no-commit gh-USER/BRANCH` instead of a plain `git merge` to merge the pull request.
 * Use `git commit --author "Author Name <author@email>"` to commit the merge with a very descriptive commit message:
    * Fill in  `Reviewed by: Your Name <FIXME>`
    * Link to a pull request `https://github.com/fenryxo/test/pull/2` → ` - Pull Request: fenryxo/test#2`
    * Link to an issue `https://github.com/fenryxo/test/issues/3` -> `+ - Issue: fenryxo/test#3`
 * Finally, ``git push``.
 * Optionally, you can remove remote repository: `git remote remove gh-USER/BRANCH`

**Comment by eurythmia** [comment]
@fenryxo I found that by following this workflow I got all of the commit messages for the development process, and by including the `git commit --author "Author Name <author@email>"` we get essentially an extra commit message for the contents of the pull request (there are all the commits that occurred during development, and the final commit message for the pull request, then there is the commit message for the merge itself).

Perhaps we should consider the following:
1) ensure all commits for the pull request have been squashed into one commit.
** this can be acheived by the requestor with a `git rebase -i git@github.com:tiliado/<repository>/master` and marking all but the first commit as `squash`
2) we can use `git merge --no-ff gh-USER/BRANCH` and stick with the default merge message. 
3) squash the pull request and merge commits so that there is only one commit, use the commit message from the pull request (as it is the descriptive message that we care about).

... using this process we can ensure there are clean, descriptive commit messages, and that the "microcommit" messages from development are no longer present. These small commits from development don't really matter to us because they are a development tool, but our commits are used for auditability of changes and to describe what is happening in any given release.

**Comment by SteVwonder** [comment]
@eurythmia, whose information did you put for the --author flag?  I believe that should be the info of the person who wrote the PR, not the person reviewing/merging the PR.
@fenryxo, that was something that tripped me up at first as well, maybe it's clearer to put `--author "Pull Request Author Name <author@email.com>"`?  Or add a sub-bullet explaining this caveat?

**Comment by fenryxo** [comment]
@eurythmia 
> I found that by following this workflow I got all of the commit messages for the development process, and by including the `git commit --author "Author Name <author@email>"` we get essentially an extra commit message for the contents of the pull request (there are all the commits that occurred during development, and the final commit message for the pull request, then there is the commit message for the merge itself).

How did you get two commit messages for "the final commit message for the pull request, then there is the commit message for the merge itself"? This should be one commit. Could you upload a sample branch somewhere?

> ensure all commits for the pull request have been squashed into one commit.
** this can be acheived by the requestor with a `git rebase -i git@github.com:tiliado/<repository>/master` and marking all but the first commit as `squash`

I would like to avoid rebasing & squashing to keep entrance barrier for new service maintainers as low as possible. As a service maintainer, you can optionally squash commits into one, but you should not request this to be done by pull requesters.

@SteVwonder, good catch! I'll fix it:-)

**Comment by fenryxo** [comment]
Updated:

#### Merge via command line

 * Add remote repository if necessary: `git remote add gh-USER https://github.com/USER/REPO.git; git fetch gh-USER`
 * Switch to the branch of the pull request: `git checkout gh-USER/BRANCH`
 * Review and test changes.
 * Switch to the master branch: `git checkout master` 
 * Use `git merge --no-ff --no-commit gh-USER/BRANCH` instead of a plain `git merge` to merge the pull request.
 * Use `git commit --author "Pull Request Author Name <author@email>"` to commit the merge with a very descriptive commit message:
    * Fill in  `Reviewed by: Your Name <FIXME>`
    * Link to a pull request `https://github.com/fenryxo/test/pull/2` → ` - Pull Request: fenryxo/test#2`
    * Link to an issue `https://github.com/fenryxo/test/issues/3` -> `+ - Issue: fenryxo/test#3`
 * Optionally, you can squash all commits of the pull request and the merge commit into one final commit. However, you should not ask author of the pull request to do rebasing and squashing to keep entrance barrier for contributors as low as possible.
 * Finally, ``git push``.
 * Optionally, you can remove remote repository: `git remote remove gh-USER/BRANCH`

**Comment by fenryxo** [comment]
* [Git & GitHub guidelines have been added to developer documentatioc](http://tiliado.github.io/nuvolaplayer/development/apps/guidelines.html#git-github-guidelines)
* [CONTRIBUTING.md template is ready](https://github.com/tiliado/nuvolaplayer/blob/master/web_apps/template/CONTRIBUTING.md)

TODO: Create pull requests with CONTRIBUTING.md

**Comment by fenryxo** [comment]
I'm closing this ticket in favour of #242, which track the status of the CONTRIBUTING.md file in all scripts.
